Cython/Compiler/Nodes.py

```diff
@@ -4731,8 +4725,152 @@ def generate_execution_code(self, code):
         code.mark_pos(self.pos)
         if self.body:
             self.body.generate_execution_code(code)
+        if not self.entry.type.early_init:
+            if self.type_init_args:
+              self.type_init_args.generate_evaluation_code(code)
+              bases = "PyTuple_GET_ITEM(%s, 1)" % self.type_init_args.result()
+              first_base = "((PyTypeObject*)PyTuple_GET_ITEM(%s, 0))" % bases
+              # Let Python do the base types compatibility checking.
+              trial_type = code.funcstate.allocate_temp(PyrexTypes.py_object_type, True)
+              code.putln("%s = PyType_Type.tp_new(&PyType_Type, %s, NULL);" % (
+                  trial_type, self.type_init_args.result()))
+              code.putln(code.error_goto_if_null(trial_type, self.pos))
+              code.put_gotref(trial_type)
+              code.putln("if (((PyTypeObject*) %s)->tp_base != %s) {" % (
+                  trial_type, first_base))
+              code.putln("PyErr_Format(PyExc_TypeError, \"best base '%s' must be equal to first base '%s'\",")
+              code.putln("             ((PyTypeObject*) %s)->tp_base->tp_name, %s->tp_name);" % (
+                         trial_type, first_base))
+              code.putln(code.error_goto(self.pos))
+              code.putln("}")
+              code.funcstate.release_temp(trial_type)
+              code.put_incref(bases, PyrexTypes.py_object_type)
+              code.put_giveref(bases)
+              code.putln("%s.tp_bases = %s;" % (self.entry.type.typeobj_cname, bases))
+              code.put_decref_clear(trial_type, PyrexTypes.py_object_type)
+              self.type_init_args.generate_disposal_code(code)
+              self.type_init_args.free_temps(code)
+
+            self.generate_type_ready_code(self.entry, code, True)
+
+    # Also called from ModuleNode for early init types.
+    @staticmethod
+    def generate_type_ready_code(entry, code, heap_type_bases=False):
+        # Generate a call to PyType_Ready for an extension
+        # type defined in this module.
+        type = entry.type
+        typeobj_cname = type.typeobj_cname
+        scope = type.scope
+        if scope: # could be None if there was an error
```

Comment:
I know, the code was like this before, so this comment is really misplaced here (sorry), but now that this old code is touched anyway, let me say that it would be a good opportunity to clean up this kind of useless level of indentation. If the code handles an error case, it should just say so and bail out early, instead of using lengthy indented if-blocks without an else. I don't mind long functions at all, as long as their structure is simple. But indented conditional blocks increase the complexity because if I can't see where an if-block ends, I have to assume that there is an else-case (and mentally prepare for it). If there is no else-case after a long block, then the if-block probably shouldn't be there in the first place, but be replaced by a short error handling case of the reversed condition, or be moved into a separate method entirely.

Reply:
Sure.